Runtime/Scripts/PersistentConnection.cs

```diff
@@ -0,0 +1,465 @@
+          }
+        }
+      }
+    }
+
+    async Task RespondToLatencyRequest()
+    {
+      string host = integration.GetHost();
+      Loc location = LocationService.RetrieveLocation();
+      if (hasTCPPorts)
+      {
+        Logger.Log("Sending Latency Samples (TCP) as a response to LatencyRequest from the server");
+        await integration.matchingEngine.EdgeEventsConnection.TestConnectAndPostLatencyUpdate(host, (uint)config.latencyTestPort, location);
+      }
+      else
+      {
+        Logger.Log("Sending Latency Samples (UDP) as a response to LatencyRequest from the server");
+        await integration.matchingEngine.EdgeEventsConnection.TestPingAndPostLatencyUpdate(host, location);
+      }
+    }
+
+    async void ProcessLatency(Statistics stats)
+    {
+      FindCloudletReply currentCloudlet = integration.latestFindCloudletReply;
+      if (stats.Avg < config.latencyThresholdTriggerMs)
+      {
+        FindCloudletMode prevMode = integration.Mode;
+        integration.UseFindCloudletPerformanceMode(true);
+        bool fcResult = await integration.FindCloudlet();
+        if (fcResult)
+        {
+          if (!integration.latestFindCloudletReply.Equals(currentCloudlet))
+          {
+            CompareLatencies(stats, currentCloudlet);
+            integration.mode = prevMode;
+          }
+          else
+          {
+            PropagateError(FindCloudletEventTrigger.LatencyTooHigh,
+              "New Cloudlet obtained from FindCloudletPerformanceMode is the same as old cloudlet");
+          }
+        }
+        else
+        {
+          PropagateError(FindCloudletEventTrigger.LatencyTooHigh, "FindCloudletPerformanceMode failed");
+        }
+      }
+    }
+
+    async void CompareLatencies(Statistics receivedStats, FindCloudletReply originalFindCloudlet)
+    {
+      AppPort appPort;
+      Site site;
+      var netTest = new NetTest(integration.matchingEngine);
+      if (hasTCPPorts)
+      {
+        appPort = integration.GetAppPort(LProto.Tcp, config.latencyTestPort == 0 ? 0 : config.latencyTestPort);
+        site = new Site { host = appPort.FqdnPrefix + integration.latestFindCloudletReply.Fqdn, port = appPort.PublicPort, testType = TestType.CONNECT };
+      }
+      else
+      {
+        appPort = integration.GetAppPort(LProto.Udp);
+        site = new Site { host = appPort.FqdnPrefix + integration.latestFindCloudletReply.Fqdn, port = appPort.PublicPort, testType = TestType.PING };
+      }
+      await netTest.TestSite(site);
+      double normalizedLatency = receivedStats.Avg - (receivedStats.Avg * config.performanceSwitchMargin);
+      if (site.average < normalizedLatency)
+      {
+        PropagateSuccess(FindCloudletEventTrigger.LatencyTooHigh, integration.latestFindCloudletReply);
+      }
+      else
+      {
+        //switch back to the previous cloudlet
+        integration.latestFindCloudletReply = originalFindCloudlet;
```

Comment:
@franklin-huang-mobiledgex  Shift to Previous Cloudlet if the newCloudlet doesn't have better latency.